tokio/src/scope/cancellation_token.rs

```diff
@@ -0,0 +1,1186 @@
+    /// The task was added to the wait queue at the [`CancellationToken`].
+    Waiting,
+    /// The task has been polled to completion.
+    Done,
+}
+
+/// Tracks the WaitForCancellationFuture waiting state.
+/// Access to this struct is synchronized through the mutex in the CancellationToken.
+struct WaitQueueEntry {
+    /// The task handle of the waiting task
+    task: Option<Waker>,
+    // Current polling state. This state is only updated inside the Mutex of
+    // the CancellationToken.
+    state: PollState,
+}
+
+impl WaitQueueEntry {
+    /// Creates a new WaitQueueEntry
+    fn new() -> WaitQueueEntry {
+        WaitQueueEntry {
+            task: None,
+            state: PollState::New,
+        }
+    }
+}
+
+struct SynchronizedState {
+    waiters: LinkedList<WaitQueueEntry>,
+    first_child: Option<NonNull<CancellationTokenState>>,
+    is_cancelled: bool,
+}
+
+impl SynchronizedState {
+    fn new() -> Self {
+        Self {
+            waiters: LinkedList::new(),
+            first_child: None,
+            is_cancelled: false,
+        }
+    }
+}
+
+/// Information embedded in child tokens which is synchronized through the Mutex
+/// in their parent.
+struct SynchronizedThroughParent {
+    next_peer: Option<NonNull<CancellationTokenState>>,
+    prev_peer: Option<NonNull<CancellationTokenState>>,
+}
+
+/// Possible states of a `CancellationToken`
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum CancellationState {
+    NotCancelled = 0,
+    Cancelling = 1,
+    Cancelled = 2,
+}
+
+impl CancellationState {
+    fn pack(self) -> usize {
+        self as usize
+    }
+
+    fn unpack(value: usize) -> Self {
+        match value {
+            0 => CancellationState::NotCancelled,
+            1 => CancellationState::Cancelling,
+            2 => CancellationState::Cancelled,
+            _ => unreachable!("Invalid value"),
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+struct StateSnapshot {
+    /// The amount of references to this particular CancellationToken.
+    /// `CancellationToken` structs hold these references to a `CancellationTokenState`.
+    /// Also the state is referenced by the state of each child.
+    refcount: usize,
+    /// Whether the state is still referenced by it's parent and can therefore
+    /// not be freed.
+    has_parent_ref: bool,
+    /// Whether the token is cancelled
+    cancel_state: CancellationState,
+}
+
+impl StateSnapshot {
+    /// Packs the snapshot into a `usize`
+    fn pack(self) -> usize {
+        self.refcount << 3 | if self.has_parent_ref { 4 } else { 0 } | self.cancel_state.pack()
+    }
+
+    /// Unpacks the snapshot from a `usize`
+    fn unpack(value: usize) -> Self {
+        let refcount = value >> 3;
+        let has_parent_ref = value & 4 != 0;
+        let cancel_state = CancellationState::unpack(value & 0x03);
+
+        StateSnapshot {
+            refcount,
+            has_parent_ref,
+            cancel_state,
+        }
+    }
```

Comment:
To be honest: Myself gets more confused by helpers for those basics than I find them helpful. 

And as long as all the code which uses something is close together (like in this functions) I also prefer values compared to constants, since I don't have to look up the constant before understanding what it contains. No preference on the representation of `4` however.